Optimize WeightedSumOp for two inputs

xiaomengy opened the pull request:
Summary: Optimize WeightedSumOp for four inputs (2 weighted inputs), just call one kernel instead of two kernels.

Differential Revision: D9566692

BIT-silence commented:
@pytrochbot retest this please

BIT-silence commented:
Some benchmark on GPU

input_shape = [100, 100, 100, 500],
param = model.WeightedSum([param, One, gradient, lr], param)
orginal: 46.43ms
new CUDA: 32.15ms
new CUDNN: 31.16ms

So this diff can reduce about 30% time of WeightedSumOp on training jobs.

houseroad commented on caffe2/operators/utility_ops.h:
```diff
@@ -321,45 +321,70 @@ class WeightedSumOp : public Operator<Context> {
-    for (int i = 2; i < InputSize(); i += 2) {
-      auto& X = Input(i);
+    for (int i = 4; i < input_size; i += 2) {
+      const auto& Xi = Input(i);
```
good name :-)

houseroad reviewed (approved):
LGTM

BIT-silence commented:
@pytorchbot retest this please

xiaomengy committed:
Optimize WeightedSumOp for two inputs (#11049)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11049

Optimize WeightedSumOp for two inputs

Reviewed By: houseroad

Differential Revision: D9566692

fbshipit-source-id: d14d2b3105947f99a780848f4317acc3fb717572
